### PR TITLE
Pools: Meta stable pool rates

### PR DESCRIPTION
**Pull request by facuspagnuolo**

### CODE REVIEW

File: pkg/pool-stable/test/meta/MetaStablePool.test.ts [hunk 1]
```diff
@@ -534,4 +535,74 @@ describe('MetaStablePool', function () {
...
+          rateProviders[0].mockRate(fp(1).add(fp(delta)));
+          rateProviders[1].mockRate(fp(1).add(fp(delta * 2)));
+        });
+      };
+
+      const itAdaptsTheScalingFactorsCorrectly = () => {
+        it('adapt the scaling factors with the price rate', async () => {
+          const priceRates = await Promise.all(rateProviders.map((provider) => provider.getRate()));
+
+          const scalingFactors = await pool.instance.getScalingFactors();
```
Comment (nventuro) [inline pkg/pool-stable/test/meta/MetaStablePool.test.ts:559]: I'd be nice to have a test as well where the scaling factors are not 1.

File: pkg/pool-stable/contracts/meta/MetaStablePool.sol [hunk 2]
```diff
@@ -17,23 +17,30 @@ pragma experimental ABIEncoderV2;
...
 import "../StablePool.sol";
 import "./OracleMiscData.sol";
 import "./StableOracleMath.sol";
 
 contract MetaStablePool is StablePool, StableOracleMath, PoolPriceOracle, IPriceOracle {
+    using FixedPoint for uint256;
     using OracleMiscData for bytes32;
 
+    address private immutable _rateProvider0;
+    address private immutable _rateProvider1;
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply pkg/pool-stable/contracts/meta/MetaStablePool.sol:34]: I decided not to use the interface there as we may could not have rate providers and it's a bit annoying to use `IRateProvider(0)` in that case

File: pkg/pool-stable/contracts/meta/MetaStablePool.sol [hunk 3]
```diff
@@ -389,4 +400,49 @@ contract MetaStablePool is StablePool, StableOracleMath, PoolPriceOracle, IPrice
...
+        return baseScalingFactor.mulDown(priceRate);
+    }
+
+    /**
+     * @dev Overrides scaling factor getter to introduce the tokens' price rate
+     */
+    function _scalingFactors() internal view virtual override returns (uint256[] memory scalingFactors) {
+        scalingFactors = super._scalingFactors();
+        uint256[] memory priceRates = _priceRates();
+        _require(scalingFactors.length == priceRates.length, Errors.INVALID_RATES_LENGTH);
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply pkg/pool-stable/contracts/meta/MetaStablePool.sol:420]: I know it shouldn't be reached, we can remove it if you prefer so

File: pkg/pool-stable/contracts/meta/MetaStablePool.sol [hunk 4]
```diff
@@ -389,4 +400,49 @@ contract MetaStablePool is StablePool, StableOracleMath, PoolPriceOracle, IPrice
...
+            scalingFactors[i] = scalingFactors[i].mulDown(priceRates[i]);
+        }
+    }
+
+    /**
+     * @dev Tells the list of price rates for each token. All price rates are fixed-point values with 18 decimals.
+     * In case there is no rate provider for a token it returns 1e18.
+     */
+    function _priceRate(IERC20 token) internal view virtual returns (uint256) {
+        return _getPriceRate(token == _token0 ? _rateProvider0 : _rateProvider1);
```
[the remark this answers is not in the document]
Comment (facuspagnuolo) [reply pkg/pool-stable/contracts/meta/MetaStablePool.sol:433]: This is exactly how we did for the weighted pool for 2 tokens, we could change it tho.
However, now that all the hooks can only be called by the vault, this should never happen, I'd pick one strategy, either commenting why we don't perform a validation or having it even if it cannot be reached.

**Review by nventuro — APPROVED** [review]

**Pull request merged**